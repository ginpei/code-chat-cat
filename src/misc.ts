import firebase from 'firebase';
import { createBrowserHistory } from 'history';

export const appHistory = createBrowserHistory();

export interface IRecord {
  createdAt: firebase.firestore.Timestamp;
  updatedAt: firebase.firestore.Timestamp;
  modelVersion: string;
}

export type ClientRecord<T> = {
  [P in Exclude<keyof T, 'modelVersion'>]: T[P];
} & { id: string; };

export type Migrations = { [key in string]: (v: IRecord) => IRecord };

export function isRecord (record: any): record is IRecord {
  return 'createdAt' in record && 'updatedAt' in record && 'modelVersion' in record;
}

export default function migrate (migrations: Migrations, record: any) {
  const startedAt = Date.now();
  while (true) {
    // most likely infinite loop
    if (Date.now() - startedAt > 3000) {
      throw new Error('Timed out');
    }

    if (!isRecord(record)) {
      console.warn(record);
      throw new Error('Invalid record data');
    }

    const { modelVersion } = record;
    const proceed = migrations[modelVersion];
    if (!proceed) {
      throw new Error(`Migration not found: ${modelVersion}`);
    }

    const result = proceed(record);
    if (result.modelVersion === record.modelVersion) {
      return result;
    }

    record = result;
  }
}
