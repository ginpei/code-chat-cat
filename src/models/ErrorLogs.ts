import firebase from '../middleware/firebase';

export type AppError =
  | Error
  | firebase.auth.Error;

export interface ErrorLog {
  error: AppError;
  id: string;
  occurredAt: number;
}

function createLog (error: AppError): ErrorLog {
  const occurredAt = Date.now();
  const random = String(Math.random()).slice(2);
  const id = `${occurredAt}${random}`;
  return { error, id, occurredAt };
}

// ----------------------------------------------------------------------------
// actions

interface AddAction {
  log: ErrorLog;
  type: 'ErrorLogs/add';
}

export function add (location: string, error: AppError): AddAction {
  console.error(`[${location}]`, error); // TODO
  return {
    log: createLog(error),
    type: 'ErrorLogs/add',
  };
}

interface ClearAction {
  type: 'ErrorLogs/clear';
}

export function clear (): ClearAction {
  return {
    type: 'ErrorLogs/clear',
  };
}

export type ErrorsAction =
  | AddAction
  | ClearAction;

// ----------------------------------------------------------------------------
// reducers

export function reduceErrorLogs (
  state: ErrorLog[] = [],
  action: ErrorsAction,
): ErrorLog[] {
  switch (action.type) {
    case 'ErrorLogs/add': {
      const logs = [...state];
      logs.push(action.log);
      return logs;
    }
    case 'ErrorLogs/clear':
      return [];
    default:
      return state;
  }
}
