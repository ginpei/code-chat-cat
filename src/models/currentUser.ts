import { Store } from 'redux';
import firebase from '../middleware/firebase';
import { CurrentUserActionTypes } from '../reducers/currentUser';

export function initialize (store: Store) {
  return new Promise<firebase.User | null>((resolve, reject) => {
    firebase.auth().onAuthStateChanged(
      (user) => {
        store.dispatch({ type: CurrentUserActionTypes.setUser, user });
        resolve(user);
      },
      (error) => reject(error),
    );
  });
}

export function logIn () {
  const email = 'test@example.com';
  const password = 'pass123';
  return firebase.auth().signInWithEmailAndPassword(email, password);
}

export function logOut () {
  return firebase.auth().signOut();
}
