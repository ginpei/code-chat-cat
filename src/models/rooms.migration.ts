import migrate, { Migrations } from '../misc';

const migrations: Migrations = {
  // example
  20190222214600: (v) => ({
    ...v,
    migrated: true,
    modelVersion: '20190222214609',
  }),
  20190222214609: (v) => v,
};

export const roomVersion = '20190222214609';

export default function migrateRoom (room: any): any {
  return migrate(migrations, room);
}
