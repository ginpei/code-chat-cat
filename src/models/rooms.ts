import firebase from '../middleware/firebase';
import { Store } from '../reducers';
import { IRoom, IRoomRecord, RoomsActionTypes, RoomStatus } from '../reducers/rooms';
import migrateRoom, { roomVersion } from './rooms.migration';

const roomsRef = firebase.firestore().collection('/rooms');

let unsubscribeUserRooms: (() => void) | null = null;
export function connectUserRooms (store: Store) {
  if (unsubscribeUserRooms) {
    return unsubscribeUserRooms;
  }
  unsubscribeUserRooms = null;

  let userId: string = '.'; // this ID never exists
  let unsubscribeUserRoomsSnapshot: () => void = () => undefined;
  const unsubscribeStore = store.subscribe(() => {
    const state = store.getState();
    const newUserId = state.currentUser.uid;
    if (newUserId === userId) {
      return;
    }
    unsubscribeUserRoomsSnapshot();
    userId = newUserId;

    if (!userId) {
      store.dispatch({
        rooms: [],
        type: RoomsActionTypes.setUserRooms,
      });
      store.dispatch({
        ready: true,
        type: RoomsActionTypes.setReady,
      });
      return;
    }

    store.dispatch({
      ready: false,
      type: RoomsActionTypes.setReady,
    });

    const userRoomsRef = roomsRef.where('userId', '==', userId);
    unsubscribeUserRoomsSnapshot = userRoomsRef.onSnapshot({
      error: (error) => {
        console.error('Error on user rooms connection', error);
      },
      next: (snapshot) => {
        const rooms = snapshot.docs
          .map((v) => snapshotToRoom(v))
          .filter((v) => v) as IRoom[];
        store.dispatch({
          rooms,
          type: RoomsActionTypes.setUserRooms,
        });
        store.dispatch({
          ready: true,
          type: RoomsActionTypes.setReady,
        });
      },
    });
  });

  unsubscribeUserRooms = () => {
    unsubscribeUserRooms = null;
    unsubscribeStore();
    unsubscribeUserRoomsSnapshot();
  };
  return unsubscribeUserRooms;
}

let unsubscribeActiveRooms: (() => void) | null = null;
export function connectActiveRooms (store: Store, callback?: () => void) {
  if (unsubscribeActiveRooms) {
    if (callback) {
      callback();
    }
    return unsubscribeActiveRooms;
  }
  unsubscribeActiveRooms = null;
  let calledBack = false;

  let unsubscribeSnapshot: () => void = () => undefined;
  const userRoomsRef = roomsRef.where('status', '==', RoomStatus.active);
  unsubscribeSnapshot = userRoomsRef.onSnapshot({
    error: (error) => {
      console.error('Error on user rooms connection', error);
    },
    next: (snapshot) => {
      const rooms = snapshot.docs
        .map((v) => snapshotToRoom(v))
        .filter((v) => v) as IRoom[];
      store.dispatch({
        rooms,
        type: RoomsActionTypes.setActiveRooms,
      });

      if (callback && !calledBack) {
        callback();
        calledBack = true;
      }
    },
  });

  unsubscribeActiveRooms = () => {
    unsubscribeActiveRooms = null;
    unsubscribeSnapshot();
  };
  return unsubscribeActiveRooms;
}

export function createRoom (
  roomData: { name: string; userId: string; },
) {
  return new Promise<IRoom>(async (resolve, reject) => {
    try {
      const record: IRoomRecord = {
        createdAt: firebase.firestore.Timestamp.now(),
        modelVersion: roomVersion,
        name: roomData.name,
        status: RoomStatus.draft,
        textbookContent: `# ${roomData.name}`,
        updatedAt: firebase.firestore.Timestamp.now(),
        userId: roomData.userId,
      };
      const docRef = await roomsRef.add(record);

      const snapshot = await docRef.get();
      const room = snapshotToRoom(snapshot);
      if (!room) {
        throw new Error('Failed to get room');
      }
      resolve(room);
    } catch (error) {
      reject(error);
    }
  });
}

export async function updateRoom (room: IRoom): Promise<void> {
  const record = roomToRecord(room);
  record.createdAt = record.createdAt || firebase.firestore.Timestamp.now();
  record.updatedAt = firebase.firestore.Timestamp.now();
  return roomsRef.doc(room.id).update(record);
}

export async function deleteRoom (room: IRoom) {
  return roomsRef.doc(room.id).delete();
}

function snapshotToRoom (snapshot: firebase.firestore.DocumentSnapshot): IRoom | null {
  const data = snapshot.data();
  if (!data) {
    return null;
  }

  const migratedData = migrateRoom(data);

  return {
    createdAt: migratedData.createdAt,
    id: snapshot.id,
    name: migratedData.name,
    status: migratedData.status,
    textbookContent: migratedData.textbookContent,
    updatedAt: migratedData.updatedAt,
    userId: migratedData.userId,
  };
}

function roomToRecord (room: IRoom): IRoomRecord {
  return {
    ...room,
    modelVersion: roomVersion,
  };
}
