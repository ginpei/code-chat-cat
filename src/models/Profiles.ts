import { useEffect, useState } from 'react';
import firebase from '../middleware/firebase';
import { noop } from '../misc';

const collectionName = 'profiles';

export function useProfile(
  auth: firebase.auth.Auth,
  firestore: firebase.firestore.Firestore,
): [Profile | null, boolean, Error | firebase.auth.Error | null] {
  const [error, setError] = useState<Error | firebase.auth.Error | null>(null);
  const [initialized, setInitialized] = useState(false);
  const [profile, setProfile] = useState<Profile | null>(null);
  const [user, setUser] = useState<firebase.User | null>(null);
  const [userInitialized, setUserInitialized] = useState(false);

  // user
  useEffect(() => auth.onAuthStateChanged(
    (currentUser) => {
      setUser(currentUser);
      setUserInitialized(true);
    },
    (err) => {
      setError(err);
      setUserInitialized(true);
      setInitialized(true);
    },
  ), [auth]);

  useEffect(() => {
    if (!userInitialized) {
      return noop;
    }

    if (!user) {
      setInitialized(true);
      return noop;
    }

    return getColl(firestore).doc(user.uid).onSnapshot(
      (ss) => {
        const result = snapshotToProfile(ss);
        setProfile(result);
        setInitialized(true);
      },
      (err) => {
        setError(err);
        setInitialized(true);
      },
    );
  }, [user, firestore, userInitialized]);

  return [profile, initialized, error];
}

export async function createNewProfile(
  firestore: firebase.firestore.Firestore,
  profile: Profile,
) {
  await getColl(firestore).doc(profile.id).set(profile);
}

export async function saveProfile2(
  firestore: firebase.firestore.Firestore,
  profile: Profile,
) {
  await getColl(firestore).doc(profile.id).set(profile);
}

function getColl(firestore: firebase.firestore.Firestore) {
  return firestore.collection(collectionName);
}

// ----------------------------------------------------------------------------
// states

export interface Profile {
  id: string;
  message?: string;
  name: string;
  type: ProfileType;
}

export enum ProfileType {
  'anonymous',
  'loggedIn',
}

export const emptyProfile: Readonly<Profile> = Object.freeze({
  id: '',
  message: '',
  name: '',
  type: ProfileType.anonymous,
});

export function getInitialProfile (userId: string): Profile {
  return {
    ...emptyProfile,
    id: userId,
    name: 'No name',
  };
}

// ----------------------------------------------------------------------------
// connectors

export function connectProfile (
  userId: string,
  onNext: (profile: Profile) => void,
  onError: (error: Error) => void = noop,
  onEach: () => void = noop,
): () => void {
  if (!userId) {
    onEach();
    return noop;
  }

  const userRef = firebase.firestore().collection(collectionName).doc(userId);
  const unsubscribeNotes = userRef.onSnapshot(
    (snapshot) => {
      const profile = snapshotToProfile(snapshot) || getInitialProfile(userId);
      onNext(profile);
      onEach();
    },
    (error) => {
      onError(error);
      onEach();
    },
  );
  return unsubscribeNotes;
}

export function saveProfile (profile: Profile) {
  const userRef = firebase.firestore().collection(collectionName).doc(profile.id);
  return userRef.set(profile);
}

export function snapshotToProfile (
  snapshot: firebase.firestore.DocumentSnapshot,
): Profile | null {
  const data = snapshot.data();
  if (!data) {
    return null;
  }

  return {
    id: data.id || '',
    message: data.message || '',
    name: data.name || '',
    type: data.type in ProfileType ? data.type : ProfileType.anonymous,
  };
}
