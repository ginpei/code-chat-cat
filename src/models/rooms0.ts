import firebase from '../middleware/firebase';
import { IRoom, IRoomRecord } from '../reducers/rooms';
import migrateRoom, { roomVersion } from './rooms.migration';

const roomsRef = firebase.firestore().collection('/rooms');

// export function createRoom (
//   roomData: { name: string; userId: string; },
// ) {
//   return new Promise<IRoom>(async (resolve, reject) => {
//     try {
//       const record: IRoomRecord = {
//         createdAt: firebase.firestore.Timestamp.now(),
//         modelVersion: roomVersion,
//         name: roomData.name,
//         status: RoomStatus.draft,
//         textbookContent: `# ${roomData.name}`,
//         updatedAt: firebase.firestore.Timestamp.now(),
//         userId: roomData.userId,
//       };
//       const docRef = await roomsRef.add(record);

//       const snapshot = await docRef.get();
//       const room = snapshotToRoom(snapshot);
//       if (!room) {
//         throw new Error('Failed to get room');
//       }
//       resolve(room);
//     } catch (error) {
//       reject(error);
//     }
//   });
// }

// export async function updateRoom (room: IRoom): Promise<void> {
//   const record = roomToRecord(room);
//   record.createdAt = record.createdAt || firebase.firestore.Timestamp.now();
//   record.updatedAt = firebase.firestore.Timestamp.now();
//   return roomsRef.doc(room.id).update(record);
// }

export async function deleteRoom (room: IRoom) {
  return roomsRef.doc(room.id).delete();
}

function snapshotToRoom (snapshot: firebase.firestore.DocumentSnapshot): IRoom | null {
  const data = snapshot.data();
  if (!data) {
    return null;
  }

  const migratedData = migrateRoom(data);

  return {
    createdAt: migratedData.createdAt,
    id: snapshot.id,
    name: migratedData.name,
    status: migratedData.status,
    textbookContent: migratedData.textbookContent,
    updatedAt: migratedData.updatedAt,
    userId: migratedData.userId,
  };
}

function roomToRecord (room: IRoom): IRoomRecord {
  return {
    ...room,
    modelVersion: roomVersion,
  };
}
