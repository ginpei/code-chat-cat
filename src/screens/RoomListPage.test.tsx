import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { noop } from '../misc';
import { emptyRoom, RoomStatus } from '../models/Rooms';
import ConnectedRoomListPage from './RoomListPage';

const RoomListPage = ConnectedRoomListPage.WrappedComponent;

describe('<RoomListPage>', () => {
  let wrapper: ShallowWrapper;
  let connectUserRooms: jest.Mock<() => void, [string]>;
  let disconnectUserRooms: jest.Mock<void, []>;

  beforeEach(() => {
    disconnectUserRooms = jest.fn(noop);
    connectUserRooms = jest.fn((_: string) => disconnectUserRooms);
  });

  describe('without any rooms', () => {
    it('just show a message', () => {
      wrapper = shallow((
        <RoomListPage
          connectUserRooms={connectUserRooms}
          userId={''}
          userRooms={[]}
        />
      ));
      expect(wrapper.text()).toMatch('No rooms found.');
    });
  });

  describe('with rooms', () => {
    beforeEach(() => {
      wrapper = shallow((
        <RoomListPage
          connectUserRooms={connectUserRooms}
          userId={''}
          userRooms={[
            {
              ...emptyRoom,
              name: 'Draft room',
              status: RoomStatus.draft,
            },
            {
              ...emptyRoom,
              name: 'Public room',
              status: RoomStatus.public,
            },
            {
              ...emptyRoom,
              name: 'Active room',
              status: RoomStatus.active,
            },
          ]}
        />
      ));
    });

    it('renders each row', () => {
      expect(wrapper.find('RoomItem')).toHaveLength(3);
    });

    it('matches to snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('rooms connestion', () => {
    beforeEach(() => {
      wrapper = shallow((
        <RoomListPage
          connectUserRooms={connectUserRooms}
          userId={''}
          userRooms={[]}
        />
      ));
    });

    it('connects to user rooms', () => {
      expect(connectUserRooms).toBeCalled();
    });

    it('disconnects when unmount', () => {
      wrapper.unmount();
      expect(disconnectUserRooms).toBeCalled();
    });
  });
});
