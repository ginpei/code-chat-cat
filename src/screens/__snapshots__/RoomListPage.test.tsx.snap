// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomListPage> with rooms matches to snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RoomListPage
    connectUserRooms={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": [MockFunction],
          },
        ],
      }
    }
    userId=""
    userRooms={
      Array [
        Object {
          "createdAt": Object {
            "toDate": [Function],
          },
          "id": "",
          "name": "Draft room",
          "status": 0,
          "textbookContent": "",
          "updatedAt": Object {
            "toDate": [Function],
          },
          "userId": "",
        },
        Object {
          "createdAt": Object {
            "toDate": [Function],
          },
          "id": "",
          "name": "Public room",
          "status": 1,
          "textbookContent": "",
          "updatedAt": Object {
            "toDate": [Function],
          },
          "userId": "",
        },
        Object {
          "createdAt": Object {
            "toDate": [Function],
          },
          "id": "",
          "name": "Active room",
          "status": 2,
          "textbookContent": "",
          "updatedAt": Object {
            "toDate": [Function],
          },
          "userId": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h1>
          Room list
        </h1>,
        <p>
          <Link
            to="/rooms/new"
          >
            Create new room
          </Link>
        </p>,
        <p>
          <Emoji
            label="Lock"
          />
           Draft / 
          <Emoji
            label="White Heavy Check Mark"
          />
           Public room / 
          <Emoji
            label="Fire"
          />
           Active room
        </p>,
        <table>
          <tbody>
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Draft room",
                  "status": 0,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Public room",
                  "status": 1,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Active room",
                  "status": 2,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
          </tbody>
        </table>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Room list",
        },
        "ref": null,
        "rendered": "Room list",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Link
            to="/rooms/new"
          >
            Create new room
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Create new room",
            "to": "/rooms/new",
          },
          "ref": null,
          "rendered": "Create new room",
          "type": [Function],
        },
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Emoji
              label="Lock"
            />,
            " Draft / ",
            <Emoji
              label="White Heavy Check Mark"
            />,
            " Public room / ",
            <Emoji
              label="Fire"
            />,
            " Active room",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "label": "Lock",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " Draft / ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "label": "White Heavy Check Mark",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " Public room / ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "label": "Fire",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " Active room",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tbody>
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Draft room",
                  "status": 0,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Public room",
                  "status": 1,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
            <RoomListItem
              room={
                Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Active room",
                  "status": 2,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                }
              }
            />
          </tbody>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Draft room",
                    "status": 0,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />,
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Public room",
                    "status": 1,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />,
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Active room",
                    "status": 2,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "",
              "nodeType": "function",
              "props": Object {
                "room": Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Draft room",
                  "status": 0,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "",
              "nodeType": "function",
              "props": Object {
                "room": Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Public room",
                  "status": 1,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "",
              "nodeType": "function",
              "props": Object {
                "room": Object {
                  "createdAt": Object {
                    "toDate": [Function],
                  },
                  "id": "",
                  "name": "Active room",
                  "status": 2,
                  "textbookContent": "",
                  "updatedAt": Object {
                    "toDate": [Function],
                  },
                  "userId": "",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "tbody",
        },
        "type": "table",
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(DefaultLayout)",
      "type": [Function],
      Symbol(Symbol.hasInstance): [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h1>
            Room list
          </h1>,
          <p>
            <Link
              to="/rooms/new"
            >
              Create new room
            </Link>
          </p>,
          <p>
            <Emoji
              label="Lock"
            />
             Draft / 
            <Emoji
              label="White Heavy Check Mark"
            />
             Public room / 
            <Emoji
              label="Fire"
            />
             Active room
          </p>,
          <table>
            <tbody>
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Draft room",
                    "status": 0,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Public room",
                    "status": 1,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Active room",
                    "status": 2,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
            </tbody>
          </table>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Room list",
          },
          "ref": null,
          "rendered": "Room list",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Link
              to="/rooms/new"
            >
              Create new room
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Create new room",
              "to": "/rooms/new",
            },
            "ref": null,
            "rendered": "Create new room",
            "type": [Function],
          },
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Emoji
                label="Lock"
              />,
              " Draft / ",
              <Emoji
                label="White Heavy Check Mark"
              />,
              " Public room / ",
              <Emoji
                label="Fire"
              />,
              " Active room",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "label": "Lock",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " Draft / ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "label": "White Heavy Check Mark",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " Public room / ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "label": "Fire",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " Active room",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tbody>
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Draft room",
                    "status": 0,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Public room",
                    "status": 1,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
              <RoomListItem
                room={
                  Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Active room",
                    "status": 2,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  }
                }
              />
            </tbody>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <RoomListItem
                  room={
                    Object {
                      "createdAt": Object {
                        "toDate": [Function],
                      },
                      "id": "",
                      "name": "Draft room",
                      "status": 0,
                      "textbookContent": "",
                      "updatedAt": Object {
                        "toDate": [Function],
                      },
                      "userId": "",
                    }
                  }
                />,
                <RoomListItem
                  room={
                    Object {
                      "createdAt": Object {
                        "toDate": [Function],
                      },
                      "id": "",
                      "name": "Public room",
                      "status": 1,
                      "textbookContent": "",
                      "updatedAt": Object {
                        "toDate": [Function],
                      },
                      "userId": "",
                    }
                  }
                />,
                <RoomListItem
                  room={
                    Object {
                      "createdAt": Object {
                        "toDate": [Function],
                      },
                      "id": "",
                      "name": "Active room",
                      "status": 2,
                      "textbookContent": "",
                      "updatedAt": Object {
                        "toDate": [Function],
                      },
                      "userId": "",
                    }
                  }
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "",
                "nodeType": "function",
                "props": Object {
                  "room": Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Draft room",
                    "status": 0,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "",
                "nodeType": "function",
                "props": Object {
                  "room": Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Public room",
                    "status": 1,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "",
                "nodeType": "function",
                "props": Object {
                  "room": Object {
                    "createdAt": Object {
                      "toDate": [Function],
                    },
                    "id": "",
                    "name": "Active room",
                    "status": 2,
                    "textbookContent": "",
                    "updatedAt": Object {
                      "toDate": [Function],
                    },
                    "userId": "",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "tbody",
          },
          "type": "table",
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(DefaultLayout)",
        "type": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
